skylog/storage/cloudspanner/sequence.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudspanner
+
+import (
+	"context"
+
+	"cloud.google.com/go/spanner"
+	"github.com/google/trillian/skylog/storage"
+)
+
+// SequenceOpts configures the sequence storage sharding mechanism.
+type SequenceOpts struct {
+	BatchSize uint64
+	Shards    uint64
+}
+
+// SequenceReader allows reading from and writing to a sequence storage.
+type SequenceStorage struct {
+	c    *spanner.Client
+	id   int64
+	opts SequenceOpts
+}
+
+// NewSequenceStorage returns a new SequenceStorage.
+func NewSequenceStorage(c *spanner.Client, treeID int64, opts SequenceOpts) *SequenceStorage {
+	return &SequenceStorage{c: c, id: treeID, opts: opts}
+}
+
+// Read fetches the specified [begin, end) range of entries, and returns them
+// in order. May return a prefix of the requested range if it spans multiple
+// shards or some entries are missing.
+func (s *SequenceStorage) Read(ctx context.Context, begin, end uint64) ([]storage.Entry, error) {
+	if end <= begin { // Empty range.
+		return nil, nil
+	}
+	// TODO(pavelkalinnikov): Restrict the range length.
+	// TODO(pavelkalinnikov): Don't return tail entries if all empty.
+	ret := make([]storage.Entry, end-begin)
+
+	// We only read entries from the shard that includes the begin-th entry.
+	// TODO(pavelkalinnikov): Unit-test this logic.
+	offset := begin / s.opts.BatchSize
+	if next := (offset + 1) * s.opts.BatchSize; end < next {
```

Comment:
this looks like it'll make this method read more entries than requested?

Reply:
Yep, the condition should be reversed.